components/board/commonsysfs/ioctl.go

```diff
@@ -0,0 +1,168 @@
+		// If errno is 0, there was no error, so ignore the (lack of) problem.
+		return nil
+	}
+	return err
+}
+
+// gpioHandleRequest is a struct to give to ioctl when configuring a GPIO pin.
+type gpioHandleRequest struct {
+	LineOffsets   [64]uint32
+	Flags         uint32
+	DefaultValues [64]byte
+	ConsumerLabel [32]byte
+	Lines         uint32
+	Fd            int32
+}
+
+// gpioHandleData is a struct to give to ioctl when reading the state of an input GPIO pin.
+type gpioHandleData struct {
+	Values [64]uint8
+}
+
+type ioctlPin struct {
+	// These values should both be considered immutable. The mutex is only here so that the ioctl
+	// calls don't have race conditions.
+	devicePath string
+	offset     uint32
+	mu         sync.Mutex
+}
+
+// This helps implement the board.GPIOPin interface for ioctlPin.
+func (pin *ioctlPin) Set(ctx context.Context, isHigh bool, extra map[string]interface{}) error {
+	pin.mu.Lock()
+	defer pin.mu.Unlock()
+
+	var value byte
+	if isHigh {
+		value = 1
+	} else {
+		value = 0
+	}
+
+	devFile, err := os.Open(pin.devicePath)
+	if err != nil {
+		return err
+	}
+	defer devFile.Close()
+
+	request := gpioHandleRequest{
+		LineOffsets:   [64]uint32{pin.offset},
+		Flags:         gpioHandleRequestOutput,
+		DefaultValues: [64]byte{value},
+		ConsumerLabel: [32]byte{},
+		Lines:         1,
+		Fd:            0,
+	}
+
+	err = ioctl(devFile.Fd(), gpioGetLineHandleIoctl, uintptr(unsafe.Pointer(&request)))
+	if err != nil {
+		return err
+	}
+	syscall.Close(int(request.Fd))
+	return nil
+}
+
+// This helps implement the board.GPIOPin interface for ioctlPin.
+func (pin *ioctlPin) Get(ctx context.Context, extra map[string]interface{}) (bool, error) {
+	pin.mu.Lock()
+	defer pin.mu.Unlock()
+
+	devFile, err := os.Open(pin.devicePath)
+	if err != nil {
+		return false, err
+	}
+	defer devFile.Close()
+
+	request := gpioHandleRequest{
+		LineOffsets:   [64]uint32{pin.offset},
+		Flags:         gpioHandleRequestInput,
+		DefaultValues: [64]byte{},
+		ConsumerLabel: [32]byte{},
+		Lines:         1,
+		Fd:            0,
+	}
+
+	err = ioctl(devFile.Fd(), gpioGetLineHandleIoctl, uintptr(unsafe.Pointer(&request)))
+	if err != nil {
+		return false, err
+	}
+	defer syscall.Close(int(request.Fd))
+
+	readRequest := gpioHandleData{Values: [64]uint8{}}
+	err = ioctl(uintptr(request.Fd), gpioHandleGetLineValuesIoctl,
+		uintptr(unsafe.Pointer(&readRequest)))
+	if err != nil {
+		return false, err
+	}
+
+	return (readRequest.Values[0] != 0), nil
+}
+
+// This helps implement the board.GPIOPin interface for ioctlPin.
+func (pin *ioctlPin) PWM(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	return 0.0, errors.New("PWM stuff is not supported on ioctl pins yet")
+}
+
+// This helps implement the board.GPIOPin interface for ioctlPin.
+func (pin *ioctlPin) SetPWM(ctx context.Context, dutyCyclePct float64, extra map[string]interface{}) error {
+	return errors.New("PWM stuff is not supported on ioctl pins yet")
```

Comment:
will it be possible to support it in ioctl?

Reply:
Eric and I discussed: the answer isn't very clear. By default, it doesn't seem like ioctl has PWM capabilities already built out, but there are folks online who have created such an interface (e.g., here's [BeagleBone instructions](https://elinux.org/BeagleBoard/GSoC/2010_Projects/Pulse_Width_Modulation)). How hard would this be to generalize to other boards? I don't know. It's outside the scope of the current ticket/PR, and Eric's question was mere curiosity and not a blocker.